Automattic/vip-go-mu-plugins
# Add integration code for block data API

## Description

This adds a `vip-integrations.php` file and two constants for enabling the [VIP Block Data API](https://github.com/Automattic/vip-block-data-api/):

- `VIP_BLOCK_DATA_API_ENABLE` (`true` or `false`, default disabled)
- This was removed per @mehmoodak's comments in https://github.com/Automattic/vip-go-mu-plugins/pull/4406/commits/b73e9dfa26899ddc89f5e12455d8548d89d48100: ~`VIP_BLOCK_DATA_API_VERSION` (pinned version number from [vip-go-mu-plugins-ext](https://github.com/Automattic/vip-go-mu-plugins-ext), defaults to latest version of `vip-block-data-api`)~

These changes go along with [this PR in -ext](https://github.com/Automattic/vip-go-mu-plugins-ext/pull/30) to add the Block Data API to the `vip-integrations/` folder.

The `PluginLoadStrategy` used is largely to ensure that mu-plugins has the ability to decide load order if a plugin is both installed manually and enabled via constant. This is set to `PREFER_CUSTOMER_CODE_VERSION` currently which means if the Block Data API plugin is manually installed, that version will override the constant-enabled version.

Currently all of the code in `vip-integrations.php` is focused on loading vip-block-data-api, but this can later be generalized to handle other external plugin loading.

## Changelog Description

### Added constant to enable the VIP Block Data API plugin

Added support to enable the [VIP Block Data API plugin](https://github.com/Automattic/vip-block-data-api/) via the `VIP_BLOCK_DATA_API_ENABLE` constant.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] N/A - This change uses a rollout method to ease with deployment.
- [x] External documentation in progress - This change has relevant documentation additions / updates (if applicable).

## Steps to Test

1. Wait until a `vip-go-mu-plugins` deployment is made, which should pull [the `vip-integrations` folder from `-ext`](https://github.com/Automattic/vip-go-mu-plugins-ext/tree/trunk/vip-integrations) into [vip-go-mu-plugins-built](https://github.com/Automattic/vip-go-mu-plugins-built).
2. Clone [vip-go-mu-plugins-built](https://github.com/Automattic/vip-go-mu-plugins-built) locally set `vip dev-env`'s environment to use it, e.g.

    ```bash
    vip dev-env update --mu-plugins=~/projects/mu-plugins-built
    ```

3. Verify the block data API plugin is disabled by default. Visit the REST endpoint for an existing post ID:

    ```bash
    GET /wp-json/vip-block-data-api/v1/posts/<post_id>/blocks

    # e.g. GET http://my-site.vipdev.lndo.site/wp-json/vip-block-data-api/v1/posts/1/blocks
    ```

    Ensure you receive a `404`response with status code `rest_no_route`.

4. In `vip-config.php`, add:

    ```php
    define( 'VIP_BLOCK_DATA_API_ENABLE', true );
    ```

6. Visit the same URL as step 3, and ensure the block data API is now available:

    ```
    GET http://my-site.vipdev.lndo.site/wp-json/vip-block-data-api/v1/posts/1/blocks

    {"blocks":[{"name":"core/paragraph","attributes":{"content":"Welcome to WordPress. This is your first post. Edit or delete it, then start writing!","dropCap":false}}]}
    ```

Follow section removed via [this conversation](https://github.com/Automattic/vip-go-mu-plugins/pull/4406#discussion_r1179942240).

<details>

<summary>(removed) The `PluginLoadStrategy` code can be tested by:</summary>

1. Disable the mu-plugins version of the block data API if enabled:

    ```php
    define( 'VIP_BLOCK_DATA_API_ENABLE', false );
    ```

2. Install [vip-block-data-api](https://github.com/Automattic/vip-block-data-api/) manually in your site's plugins and enable it.
3. Modify the locally installed version of the plugin by adding an early return in [`src/rest/rest-api.php`'s `get_block_content()` function](https://github.com/Automattic/vip-block-data-api/blob/585e000/src/rest/rest-api.php#L83):

    ```php
    public static function get_block_content( $params ) {
        return ['local_plugin' => true];
        // ...
    }
    ```

4. Retrieve a post's block content via the rest API:

    ```
    GET http://my-site.vipdev.lndo.site/wp-json/vip-block-data-api/v1/posts/1/blocks
    ```

    Ensure this shows the local plugin `{'local_plugin' => true}` response.

5. Enable the mu-plugins constant for the plugin:

    ```php
    define( 'VIP_BLOCK_DATA_API_ENABLE', true );
    ```

7. Query the REST URL from step 4. Ensure that the `{'local_plugin' => true}` response is still present. This should be the case when `$plugin_load_strategy` is set to `PREFER_CUSTOMER_CODE_VERSION`. 
8. In `vip-integrations.php`, change `$plugin_load_strategy` to `PREFER_MU_PLUGINS_VERSION`:

    ```diff
    - $plugin_load_strategy = PluginLoadStrategy::PREFER_CUSTOMER_CODE_VERSION;
    + $plugin_load_strategy = PluginLoadStrategy::PREFER_MU_PLUGINS_VERSION;
    ```
9. Query the REST URL from step 4 again. This time, it should use the mu-plugins version of the REST API and return the block's contents.

</details>
